### PR TITLE
Simulation save/restart

### LINKED ISSUE #129: Restart capability
ABLATE needs the ability to stop and restart a simulation.

### PR DESCRIPTION
**Pull request by mmcgurn**
Merges an existing interface into the Serializable interface and allows for simulation save/restart and closes #129 . This PR also includes some bug fixes found in prep for the review meeting.

### CODE REVIEW

File: ablateLibrary/particles/particles.cpp [hunk 1]
```diff
@@ -544,25 +544,108 @@ static PetscErrorCode DMSequenceViewTimeHDF5(DM dm, PetscViewer viewer) {
...
+    VecDestroy(&particlesPerRank) >> checkError;
+
     // if this is an hdf5Viewer
     PetscBool ishdf5;
     PetscObjectTypeCompare((PetscObject)viewer, PETSCVIEWERHDF5, &ishdf5) >> checkError;
     if (ishdf5) {
         DMSequenceViewTimeHDF5(GetDM(), viewer) >> checkError;
     }
 }
+
```
Comment (mmcgurn) [inline ablateLibrary/particles/particles.cpp:586]: @knepley Do you know if there is a better way to save/restore to dmSwarm?

Comment (knepley) [reply ablateLibrary/particles/particles.cpp:586]: Currently, no. You are right that we should move this code into DMView()/DMLoad()

**Review by knepley — APPROVED** [review]
I will have to try this out now.

File: ablateLibrary/particles/particles.cpp [hunk 2]
```diff
@@ -544,25 +544,118 @@ static PetscErrorCode DMSequenceViewTimeHDF5(DM dm, PetscViewer viewer) {
...
             DMSwarmDestroyGlobalVectorFromField(GetDM(), field.fieldName.c_str(), &particleVector) >> checkError;
         }
     }
 
+    // Get the particle info
+    int rank;
+    MPI_Comm_rank(PetscObjectComm((PetscObject)GetDM()), &rank) >> checkMpiError;
+
+    // get the local number of particles
+    PetscInt globalSize;
```
Comment (mmcgurn) [inline ablateLibrary/particles/particles.cpp:565]: @dsalac These are changes we talked about today.  I'm saving the global size of the dmSwarm and then restoring the particles "evenly" on resume. There is no assumption on save/restore ranks.

**Pull request merged**